### DISCUSSION TITLE
Typescript | making custome components the right way

### DISCUSSION THREAD

**Discussion by Emiltayeb** [chakra-ui/chakra-ui#4558] [Q&A] [answered]
### Hello everyone!  👋

I have just started using Charkra and I absolutely love it' and alsot i started learning typescript.
I was trying to understand something a bit complex.

I was thinking about creating my own Input and Alert component.

It would include props such as `hassError` and a custom status name such as `status`.

My code look somthing like this:

### Form.tsx:

```typescript
          <AlertInput
            title='Password must be at least 6 digits'
            hasError={isPasswordHasErrors}>
        // ...rest of the input element
          </AlertInput>
```

### AlertInput.tsx

```typescript

interface IProps {
  hasError: Boolean;
  status?: string;
  title: string;
}

const AlertInput: React.FC<IProps> = ({
  children,
  hasError,
  status,
  title,
}) => {
  return (
    <>
      {children}
      {hasError && (
        <Alert mb={6} status={status ?? 'error'}>
          {title}
        </Alert>
      )}
    </>
  );
};
```
My goal is to make it so that if I don't pass the `status` prop, it will default to an error.
I tried everything, even setting the default prop to `error`, and yet typescript kept screaming at me:

```Type 'string' is not assignable to type '"warning" | "error" | "info" | "success" | undefined'```

Well, it confused me becuase error is of type string. so i  add this type to  the `IProp` interface like this:

```typescript
interface IProps {
  hasError: Boolean;
  status?: 'error' | 'info' | 'warning' | 'success';
  title: string;
}
```

It's hard for me to understand why it works. Why did manually setting 'error' not work?
Can I access predefined types through chakra next time?
I may want to make a custom icon component. Which prop type should I use for chakra icons?

Thank you very much for the help 💓

**Comment by payapula** [comment, marked as answer]
Hi @Emiltayeb 

> It's hard for me to understand why it works. Why did manually setting 'error' not work?

When we are setting the type `'error' | 'info' | 'warning' | 'success'` , it by default extends `'string'` type. When we give the string `'error'` directly to chakra's status prop, which expects union, TS complains about type mismatch. More details can be found here: https://stackoverflow.com/a/51101237. 

And by the way, we can import chakra's types from the chakra package and use them in your files.

```
import { AlertStatus } from "@chakra-ui/react";

interface IProps {
  hasError: Boolean;
  status?: AlertStatus;
  title: string;
}
```

Full sandbox link - https://codesandbox.io/s/admiring-flower-40vc5?file=/src/index.tsx:170-252

> I may want to make a custom icon component. Which prop type should I use for chakra icons?

You can check what prop types that chakra's Icon component extends in source code - https://github.com/chakra-ui/chakra-ui/blob/main/packages/icon/src/icon.tsx#L0-L1

If you need help in any specific area, kindly create a sandbox from chakra's templates and post.

## CodeSandbox Templates

- JavaScript Starter: https://codesandbox.io/s/chakra-ui-javascript-lzzg9
- TypeScript Starter: https://codesandbox.io/s/chakra-ui-typescript-pomi8
- NextJS TypeScript Starter:
  https://codesandbox.io/s/chakra-ui-next-js-typescript-kxvyr

**Reply by Emiltayeb** [reply]
Hey!

thank you very much for that reply. you really made is simple to understand.

> And by the way, we can import chakra's types from the chakra package and use them in your files.

this is amazing. i tried to find this on Chakra docs but couldnt find it.
how do you know what to import?

**Reply by payapula** [reply]
You are most welcome.

> how do you know what to import?

Mostly we won't be needing this, but there are some special cases like what you mentioned.

In that case, I will check the typescript error and check chakra's source code if they are exporting the Interface/Type that the typescript is expecting. Definitely, that would be exported, and you can import them in your files.
